File: src/omnia/data/structure/rx/GenericObservableChannels.java
package omnia.data.structure.rx;

import io.reactivex.Flowable;

class GenericObservableChannels<
        StateType, MutationEventType extends ObservableDataStructure.MutationEvent>
    implements ObservableDataStructure.ObservableChannels {

  private final Flowable<StateType> states;
  private final Flowable<MutationEventType> mutations;

  protected GenericObservableChannels(
      Flowable<StateType> states, Flowable<MutationEventType> mutations) {
    this.states = states;
    this.mutations = mutations;
  }

  @Override
  public Flowable<StateType> states() {
    return states;
  }

  @Override
  public Flowable<MutationEventType> mutations() {
    return mutations;
  }
}

File: omnia/src/commonMain/omnia/data/structure/observable/writable/GenericObservableChannels.kt
package omnia.data.structure.observable.writable

import com.badoo.reaktive.observable.Observable
import omnia.data.structure.observable.ObservableDataStructure

internal open class GenericObservableChannels<
    StateType : Any,
    MutationEventType : ObservableDataStructure.MutationEvent>
protected constructor(
    private val states: Observable<StateType>,
    private val mutations: Observable<MutationEventType>) :
    ObservableDataStructure.ObservableChannels {

  override fun states(): Observable<StateType> {
    return states
  }

  override fun mutations(): Observable<MutationEventType> {
    return mutations
  }
}